File: python/ch06_agent/agent_ch06/core/commands.py
"""操作系统命令执行边界。

这是什么：命令执行的接口定义和结果对象
Java 类比：interface CommandRunner + record CommandResult
为什么需要：抽象命令执行细节，让核心代码不依赖具体的进程管理实现
"""

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True, slots=True)
class CommandResult:
    """操作系统命令执行后的统一结果。

    这是什么：命令执行结果的数据对象
    Java 类比：record CommandResult(String output, int exitCode, boolean timedOut, boolean truncated)
    为什么需要：封装命令执行的结果信息，避免直接暴露底层进程对象

    不直接返回 subprocess 对象，是为了让核心层不依赖具体操作系统 API。
    """

    output: str  # stdout 和 stderr 合并后的文本。
    exit_code: int  # 进程退出码，通常 0 表示成功。
    timed_out: bool  # 是否因为超过时间限制被终止。
    truncated: bool  # 输出是否超过上限并被截断。


class CommandRunner(Protocol):
    """命令执行接口；真实实现和测试 Fake 都遵守它。

    这是什么：命令执行器的协议接口
    Java 类比：interface CommandRunner
    为什么需要：定义命令执行的契约，支持依赖注入和测试替换

    Java 对照：这就是一个只有一个核心方法的 `interface`。AgentRunner
    只依赖这个接口，因此测试时可以传入不会真正启动 PowerShell 的 Fake。
    """

    def run(self, command: str, cwd: str, timeout_ms: int | None = None) -> CommandResult:
        """在受控工作目录中执行一条命令，并返回与操作系统无关的结果对象。

        这是什么：命令执行方法
        Java 类比：CommandResult run(String command, String cwd, Integer timeoutMs)
        为什么需要：提供统一的命令执行接口，隔离平台差异
        """

File: typescript/chapters/ch01/src/core/commands.ts
// 命令适配器向工具层暴露的最小结果契约，保留受限执行的状态信息。
//
// CommandResult 包含运行时所有可能结果状态，工具层据此决定 toolSuccess 或具体 toolError。
export interface CommandResult {
  // 合并 stdout 与 stderr 后的受限文本，供工具回填给模型。
  readonly output: string;
  // 子进程退出状态；非零值由上层转换为可恢复的工具错误。
  readonly exitCode: number;
  // 超时终止与普通非零退出需要区别处理，因此单独保留此标记。
  readonly timedOut: boolean;
  // 输出超过共享预算时为 true，调用方据此避免把不完整结果当成全部结果。
  readonly truncated: boolean;
}

//
// CommandRunner 抽象进程执行边界，使核心循环不依赖具体子进程实现；
// 测试可使用 FakeCommandRunner 注入确定性结果以隔离操作系统依赖。
//
export interface CommandRunner {
  // 命令工具依赖的可替换进程边界；实现负责 cwd、超时及输出收集。
  // timeoutMs 仅覆盖本次调用的时间预算，省略时由具体适配器使用默认限制。
  run(command: string, cwd: string, timeoutMs?: number): Promise<CommandResult>;
}
